src/hevm/src/EVM.hs

```diff
@@ -1905,12 +1908,63 @@ cheatActions =
               accessStorage a' (mksym slot) $ \res -> do
                 assign (state . returndata . word256At 0) res
                 assign (state . memory . word256At outOffset) res
+          _ -> vmError (BadCheatCode sig),
+
+      action "sign(uint256,bytes32)" $
+        \sig outOffset _ input -> case decodeStaticArgs input of
+          [sk, hash] ->
+            forceConcrete (mksym sk) $ \sk' ->
+              forceConcrete (mksym hash) $ \(C _ hash') -> let
+                curve = getCurveByName SEC_p256k1
+                priv = PrivateKey curve (num sk')
+                digest = digestFromByteString (word256Bytes hash')
+              in do
+                case digest of
+                  Nothing -> vmError (BadCheatCode sig)
+                  Just digest' -> do
+                    let s = ethsign priv digest'
+                        v = if (sign_s s) % 2 == 0 then 27 else 28
+                        encoded = encodeAbiValue $
+                          AbiTuple (RegularVector.fromList
+                            [ AbiUInt 8 v
+                            , AbiBytes 32 (encodeAbiValue . AbiUInt 256 . fromInteger $ sign_r s)
+                            , AbiBytes 32 (encodeAbiValue . AbiUInt 256 . fromInteger $ sign_s s)
+                            ])
+                    assign (state . returndata) (ConcreteBuffer encoded)
+                    copyBytesToMemory (ConcreteBuffer encoded) (num . BS.length $ encoded) 0 outOffset
+          _ -> vmError (BadCheatCode sig),
+
+      action "addr(uint256)" $
+        \sig outOffset _ input -> case decodeStaticArgs input of
+          [sk] -> forceConcrete (mksym sk) $ \sk' -> let
+                curve = getCurveByName SEC_p256k1
+                pubPoint = generateQ curve (num sk')
+                encodeInt = encodeAbiValue . AbiUInt 256 . fromInteger
+              in do
+                case pubPoint of
+                  PointO -> do vmError (BadCheatCode sig)
+                  Point x y -> do
+                    -- See yellow paper #286
+                    let
+                      pub = BS.concat [ encodeInt x, encodeInt y ]
+                      addr = w256lit . num . word256 . BS.drop 12 . BS.take 32 . keccakBytes $ pub
+                    assign (state . returndata . word256At 0) addr
+                    assign (state . memory . word256At outOffset) addr
           _ -> vmError (BadCheatCode sig)
+
     ]
   where
     action s f = (abiKeccak s, f (Just $ abiKeccak s))
     mksym x = S (Todo "abidecode" []) x
 
+-- | Hack deterministic signing, totally insecure...
+ethsign :: PrivateKey -> Digest Crypto.Keccak_256 -> Signature
+ethsign sk digest = go 420
```

Comment:
is there a chance this could loop forever?

Reply:
hmmmmmmm, I'm not sure, it's pretty much exactly the same as the implementation in `cryptonite`  ([src](https://github.com/haskell-crypto/cryptonite/blob/cryptonite-v0.27/Crypto/PubKey/ECC/ECDSA.hs#L100)), just with a monotonically increasing `k` instead of a randomly selected one...